src/backend/vulkan/src/device.rs

```diff
@@ -281,13 +281,70 @@ impl d::Device<B> for Device {
         }
     }
 
-    fn create_pipeline_cache(&self) -> Result<n::PipelineCache, d::OutOfMemory> {
+    fn create_pipeline_cache(
+        &self,
+        data: Option<Vec<u8>>,
+        device: &PhysicalDevice,
+    ) -> Result<n::PipelineCache, d::OutOfMemory> {
+        fn bad_cache(d: &[u8], properties: &vk::PhysicalDeviceProperties) -> bool {
+            let header = unsafe { std::mem::transmute::<[u8; 4], u32>([d[0], d[1], d[2], d[3]]) };
+            if header <= 0 {
+                warn!("Bad header length {:?}", header);
+                return true;
+            }
+
+            let cache_header_version = unsafe { std::mem::transmute::<[u8; 4], i32>([d[4], d[5], d[6], d[7]]) };
+            if cache_header_version != vk::PipelineCacheHeaderVersion::ONE.as_raw(){
+                warn!("Unsupported cache header version: {:?}", cache_header_version);
+                return true;
+            }
+
+            let vendor_id = unsafe { std::mem::transmute::<[u8; 4], u32>([d[8], d[9], d[10], d[11]]) };
+            if vendor_id != properties.vendor_id as u32 {
+                warn!(
+                    "Vendor ID mismatch. Device: {:?}, cache: {:?}.",
+                    properties.vendor_id,
+                    vendor_id,
+                );
+                return true;
+            }
+
+            let device_id = unsafe { std::mem::transmute::<[u8; 4], u32>([d[12], d[13], d[14], d[15]]) };
```

Comment:
looks like we should just cast the first 4 dwords right away, using `std::slice::from_raw_parts`